python/src/equistore/operations/join.py

```diff
@@ -117,118 +71,45 @@ def _join_labels(labels: List[Labels]) -> Labels:
-
-    fname = "_join_blocks_along_properties"
-    for block in blocks[1:]:
-        _check_blocks(first_block, block, ["components", "samples"], fname)
-        _check_same_gradients(first_block, block, ["components", "samples"], fname)
-
-    properties = _join_labels([block.properties for block in blocks])
-    result_block = TensorBlock(
-        values=_dispatch.concatenate([block.values for block in blocks], axis=1),
-        samples=first_block.samples,
-        components=first_block.components,
-        properties=properties,
-    )
-
-    for parameter, first_gradient in first_block.gradients():
-        # Different blocks might contain different gradient samples
-        gradient_sample_values = np.unique(
-            np.hstack([block.gradient(parameter).samples for block in blocks])
-        ).tolist()
-        gradient_samples = Labels(
-            names=first_gradient.samples.names, values=np.array(gradient_sample_values)
-        )
-
-        gradient_data = np.zeros(
-            (len(gradient_samples),)
-            + first_gradient.data.shape[1:-1]
-            + (len(properties),)
-        )
-
-        properties_start = 0
-        for block in blocks:
-            gradient = block.gradient(parameter)
-
-            # find the correct sample position to put the data
-            for i_grad, sample_grad in enumerate(gradient.samples):
-                i_sample = gradient_samples.position(sample_grad)
-                gradient_data[
-                    i_sample,
-                    ...,
-                    properties_start : properties_start + len(gradient.properties),
-                ] = gradient.data[i_grad]
-
-            properties_start += len(gradient.properties)
-
-        result_block.add_gradient(
-            parameter, gradient_data, gradient_samples, first_gradient.components
-        )
-
-    return result_block
-
+    keys_names = ("tensor",) + tensor_maps[0].keys.names
+    keys_values = []
+    blocks = []
 
-def _join_blocks_along_samples(blocks: List[TensorBlock]) -> TensorBlock:
-    """Join a sequence of :py:class:`TensorBlock` along their samples."""
-    first_block = blocks[0]
+    for i, tensor in enumerate(tensor_maps):
+        keys_values += [(i,) + value for value in tensor.keys.tolist()]
 
-    fname = "_join_blocks_along_samples"
-    for block in blocks[1:]:
-        _check_blocks(first_block, block, ["components", "properties"], fname)
-        _check_same_gradients(first_block, block, ["components", "properties"], fname)
+        for _, block in tensor:
+            samples = block.samples
+            properties = block.properties
 
-    result_block = TensorBlock(
-        values=_dispatch.concatenate([block.values for block in blocks], axis=0),
-        samples=_join_labels([block.samples for block in blocks]),
-        components=first_block.components,
-        properties=first_block.properties,
-    )
+            if not names_are_same:
+                if axis == "samples":
```

Comment:
I think `axis == "samples and not names_are_same` should be an error. It's fine to loose metadata on the property axis, less so on the sample axis

Reply:
Yes, I will raise an error and add a test.